Automattic/woocommerce-payments
# Add conditional dispute evidence fields

Fixes #283 

#### Changes proposed in this Pull Request

Adds a "Product Type" selector atop the dispute evidence form, which – along with the dispute reason (a.k.a. "category") – dynamically **controls which set of fields appears** in the form.

![duplicate](https://user-images.githubusercontent.com/1867547/76111698-bb53c900-5fae-11ea-97d6-b543da2ca020.png)

The test case data (in client/disputes/evidence/test/fields.js) spells out _all of the combinations_ (directly according to the doc in p1565734652237200-slack-wcpay), but the data structure (in client/disputes/evidence/fields.js) is _condensed to minimize repetition_. The fields that are duplicates of others are marked as `denormalized` (so that we can omit them <s>if we were</s> to have an "All fields" mode like the Stripe dashboard does).

Besides a bit of minor editing, the only difference from that doc should be that combined cards (i.e. "Cancelation policy & service information" for `subscription_canceled`) are broken out into separate ones in a couple of cases [[screenshot](https://cloudup.com/czrhPNDVNCv)].

As for the product type UI, the select control is intended to be a temporary solution here. Should we improve this before merging? (See [screenshot of Stripe's UI](https://cloudup.com/cnaqGM7tP_S).) Also, should we bother with hiding the product type selector in the rare case that it makes no difference (i.e. possibly only `credit_not_processed`)?

(See also [evidence object doc](https://stripe.com/docs/api/disputes/evidence_object).)

#### Testing instructions



Check out with a dispute test card (such as 4000000000000259), find the dispute in the Disputes list, go to details, and click "Challenge Dispute". Select a product type.

Verify that form matches expected content and structure with any given dispute reason and selected product type.

To quickly mock different reasons in the view layer, drop this snippet at bottom of client/disputes/evidence/index.js (above the `return` statement):

```js
// Use 'r' key to cycle through dispute reasons, shift+r to go backwards.
useEffect( () => {
	const reasons = [ 'bank_cannot_process', 'check_returned', 'credit_not_processed', 'customer_initiated', 'debit_not_authorized', 'duplicate', 'fraudulent', 'general', 'incorrect_account_details', 'insufficient_funds', 'product_not_received', 'product_unacceptable', 'subscription_canceled', 'unrecognized' ];
	document.addEventListener( 'keypress', ( event ) => {
		event.code === 'KeyR' && setDispute( d => {
			const index = reasons.indexOf( d && d.reason );
			const nextIndex = event.shiftKey ? index - 1 : index + 1;
			const reason = reasons[ nextIndex === reasons.length ? 0 : nextIndex === -1 ? reasons.length - 1 : nextIndex ];
			console.log( `Changing dispute reason to: ${ reason }` );
			return { ...d, reason };
		} );
	} );
}, [] );
```

-------------------
